### PR TITLE
Add code to insert/update archive folders

### PR DESCRIPTION
**Pull request by techncl**

### CODE REVIEW

File: src/test/scala/testUtils/ExternalServicesTestUtils.scala [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+import uk.gov.nationalarchives.dp.client.EntityClient.{AddEntityRequest, UpdateEntityRequest}
+import uk.gov.nationalarchives.{DADynamoDBClient, Lambda}
+
+import java.util.UUID
+import scala.collection.immutable.ListMap
+
+class ExternalServicesTestUtils extends AnyFlatSpec with BeforeAndAfterEach with BeforeAndAfterAll {
+  val graphQlServerPort = 9001
+
+  val wiremockGraphqlServer = new WireMockServer(graphQlServerPort)
```
Comment (MancunianSam) [inline src/test/scala/testUtils/ExternalServicesTestUtils.scala:26]: Are you sure graphqlServer is the right name :smile: ?

**Review by MancunianSam — COMMENTED** [review]
Looks good thanks. Just a few minor things.

File: src/main/scala/uk/gov/nationalarchives/Lambda.scala [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+
+  private val configIo: IO[Config] = ConfigSource.default.loadF[IO, Config]()
+  private implicit val secretRW: default.ReadWriter[StepFnInput] = default.macroRW[StepFnInput]
+
+  override def handleRequest(input: InputStream, output: OutputStream, context: Context): Unit = {
+    val rawInput: String = Source.fromInputStream(input).mkString
+    val stepFnInput = default.read[StepFnInput](rawInput)
+
+    val folderIdPrimaryKeysAndValues: List[PrimaryKey] =
+      stepFnInput.archiveHierarchyFolders.map(id => PrimaryKey(id))
```
[the remark this answers is not in the document]
Comment (techncl) [reply src/main/scala/uk/gov/nationalarchives/Lambda.scala:40]: Oh, cool, I didn't know you could do that with Classes. 👍

File: src/main/scala/uk/gov/nationalarchives/Lambda.scala [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+
+  private def getExpectedParentRefForEachFolder(folderIdAndInfo: Map[String, FullFolderInfo]): List[FullFolderInfo] =
+    folderIdAndInfo.map { case (_, fullFolderInfo) =>
+      val directParent = fullFolderInfo.folderRow.parentPath.split("/").last
+      folderIdAndInfo.get(directParent) match {
+        case None => fullFolderInfo // top-level folder doesn't/shouldn't have parent path
+        case Some(fullFolderInfoOfParent) =>
+          val parentEntityExists = fullFolderInfoOfParent.entity.nonEmpty
+          if (parentEntityExists)
+            fullFolderInfo.copy(expectedParentRef = fullFolderInfoOfParent.entity.get.ref.toString)
```
[the remark this answers is not in the document]
Comment (techncl) [reply src/main/scala/uk/gov/nationalarchives/Lambda.scala:163]: That's a really elegant solution. Thanks

**Review by MancunianSam — APPROVED** [review]

**Pull request merged**